oio/xcute/orchestrator.py

```diff
@@ -0,0 +1,606 @@
+            tasks_batch_size = job_config['tasks_batch_size']
+
+            tasks_run_time = 0
+            batch_per_second = tasks_per_second / float(
+                tasks_batch_size)
+            # The backend must have the tasks in order
+            # to know the last task sent
+            tasks = OrderedDict()
+            for task_id, task_payload in job_tasks:
+                if not self.running:
+                    break
+
+                tasks[task_id] = task_payload
+                if len(tasks) < tasks_batch_size:
+                    continue
+
+                tasks_run_time = ratelimit(
+                    tasks_run_time, batch_per_second)
+
+                sent = self.dispatch_tasks(
+                    beanstalkd_workers,
+                    job_id, job_type, job_config, tasks)
+                if sent:
+                    job_status, exc = self.handle_backend_errors(
+                        self.backend.update_tasks_sent, job_id, tasks.keys())
+                    tasks.clear()
+                    if exc is not None:
+                        self.logger.warn(
+                            '[job_id=%s] Sent tasks has not been updated: %s',
+                            job_id, exc)
+                        break
+                    if job_status == 'PAUSED':
+                        self.logger.info('Job %s is paused', job_id)
+                        return
+            else:
+                self.logger.info('All tasks sent (job_id=%s)' % job_id)
+
+                sent = self.dispatch_tasks(
+                    beanstalkd_workers,
+                    job_id, job_type, job_config, tasks)
+                if sent:
+                    job_status, exc = self.handle_backend_errors(
+                        self.backend.update_tasks_sent, job_id, tasks.keys(),
+                        all_tasks_sent=True)
+                    if exc is None:
+                        if job_status == 'FINISHED':
+                            self.logger.info('Job %s is finished', job_id)
+
+                        self.logger.info(
+                            'Finished dispatching job (job_id=%s)', job_id)
+                        return
+                    else:
+                        self.logger.warn(
+                            '[job_id=%s] Last sent tasks has not been '
+                            'updated: %s', job_id, exc)
+
+            _, exc = self.handle_backend_errors(self.backend.free, job_id)
+            if exc is not None:
+                self.logger.warn(
+                    '[job_id=%s] Job has not been freed: %s',
+                    job_id, exc)
+        except Exception as exc:
+            self.logger.exception(
+                '[job_id=%s] Failed generating task list: %s', job_id, exc)
+            _, exc = self.handle_backend_errors(self.backend.fail, job_id)
+            if exc is not None:
+                self.logger.warn(
+                    '[job_id=%s] Failure has not been updated: %s',
```

Comment:
Failure counter?

Reply:
For this case, there is no counter. It's all the job that fails. Therefore, we pass the status on to `FAILED`.